src/graphapi/opgraph.cpp

```
@@ -0,0 +1,109 @@
+    /// \todo might eventually move this state to the graph class during build()
+    std::unordered_map<Tensor*, EdgeInfo> e_map;
+
+    for(OpNode* n : mNodes)
+    {
+
+        for(Tensor* i : n->getInTensors())
+        {
+            auto [iter, _ignore] =
+                e_map.try_emplace(i, EdgeInfo{}); // add empty EdgeInfo if not present
+
+            iter->second.mDests.emplace_back(n);
+        }
+
+        for(Tensor* o : n->getOutTensors())
+        {
+            auto [iter, _ignore] =
+                e_map.try_emplace(o, EdgeInfo{}); // add empty EdgeInfo if not present
+
+            assert(iter->second.mSrc == nullptr);
+            iter->second.mSrc = n;
+        }
+    }
+
+    graph.initNodes(std::move(mNodes));
+
+    for(const auto& [tens_ptr, edge_info] : e_map)
+    {
+
+        if(edge_info.mSrc != nullptr && !edge_info.mDests.empty())
+        {
+            for(const auto& d : edge_info.mDests)
+            {
+                graph.addEdge(edge_info.mSrc, tens_ptr, d);
+            }
+        }
+        else if(edge_info.mSrc == nullptr)
+        {
+
+            // tens_ptr is a non-virtual input tensor
+            assert(!tens_ptr->isVirtual()); // tensor pointer can't be virtual
+            assert(!edge_info.mDests.empty());
+
+            // NOTE(Amber): we may take out this step if we decide not to add a source
+            // and sink in the graph
+            for(const auto& d : edge_info.mDests)
+            {
+                graph.addEdgeFromSrc(d, tens_ptr);
+            }
+        }
+        else if(edge_info.mDests.empty())
+        {
+            // tens_ptr is an output tensor (virtual or non-virtual)
+            assert(edge_info.mSrc != nullptr);
+            graph.addEdgeToSink(edge_info.mSrc, tens_ptr);
+        }
+        else
+        {
+            // both can't be true at the same time
+            assert(!(edge_info.mSrc == nullptr && edge_info.mDests.empty()));
```

[Quality] This condition is also checked at lines 83 and 95, which means code duplication (albeit the very local one). Please consider refactoring this.

---

This is actually "always false" `else` case.
We can get to the else case if `!(!src & !dst) & !src & !dst` which is `(src | dst) & !src & !dst` which is `(src & !src & !dst) | (dst & !src & !dst)` and as `x & !x == false` we have `false | false` `else` case.

The assert is not required at all and it's just a dead code. `Else` case is impossible.

---

`else` is possible but I moved the assert in else up to a `MIOPEN_THROW_IF`.

---

> `else` is possible but I moved the assert in else up to a `MIOPEN_THROW_IF`.

Else is not possible for given two variables.
https://godbolt.org/z/9EKnb6b15

---

[resolved]

---

ok, I see it now. The second `else if` masks the error condition in the last `else`, but now we can check for it up with the `MIOPEN_THROW_IF`.